### DISCUSSION TITLE
CommunityToolKit PopupService Public GetPopup Method

### DISCUSSION THREAD

**Discussion by bobbydharrell** [CommunityToolkit/Maui#1648] [New Feature Discussions]
I have forked this so that I could add the code in question. We use a custom PopupService in our apps that handles any popups. The need to close a popup from the ViewModel arose but we discovered there was no easy way to do this as the ViewModels had no context of the Views. I found that the Toolkit.PopupService had an internal GetPopup that would look up the Popup by ViewModel, so I added another public method that takes in ViewModel and then looks up the popup from there.

The intended use case would be from within a custom popup Service

ViewModel Calls PopupDialogService.ClosePopup<TViewModel>()

PopupDialogService.ClosePopup<TViewModel>()
{
Popup popup = PopupService.GetPopup<TViewModel>();
popup.Close();
}

I also see a need for a PopupService.OpenPopups List, but that may be another ticket, if that can be handled here then, perfect.
For clarity, the PopupService.OpenPoups would return list of all open Popups, and would be used for scenarios where one would need to close all, or see if a certain popup is currently open

See my fork for the changes
https://github.com/CommunityToolkit/Maui/compare/main...bobbydharrell:Maui:main

Thanks

**Comment by VladislavAntonyuk** [comment]
If Popup is registered as transient you will get a new instance of Popup each time.

**Reply by bobbydharrell** [reply]
You are correct, being Transient they create new popups each time. That however does not change anything about my original post.

**Reply by bijington** [reply]
I believe Vlads comment was trying to highlight that using `GetPopup` with a transient registration wouldn't solve the use case of closing an open popup because the popup that is returned won't have been shown to be closed.

If you are focussed on closing popups and not specifically opening up `GetPopup`, we do already have a feature here https://github.com/CommunityToolkit/Maui/issues/1530. Perhaps we can discuss your use cases there?

**Reply by bobbydharrell** [reply]
@bijington Thank you for the clarity, apologies @VladislavAntonyuk for not understanding you

Yes that feature could be expanded, I see it is still relying on TPopup to grab the popup and close it, We need to also allow `ClosePopup<TViewModel>` so that we can close them from a ViewModel.

But I agree, we can move this discussion to that feature. Thank you
